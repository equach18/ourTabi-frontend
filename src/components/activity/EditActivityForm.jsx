import { useState, useContext } from "react";
import { useNavigate, useParams } from "react-router-dom";
import TripContext from "../../context/TripContext";
import Alert from "../common/Alert";

function EditActivityForm() {
  const { activities, deleteActivity, editActivity } = useContext(TripContext);
  const { activityId } = useParams(); // Get activity ID from URL params
  const navigate = useNavigate();

  const validCategories = [
    "food",
    "hiking",
    "tours",
    "shopping",
    "adventure",
    "outdoors",
    "other",
  ];

  // Find the activity being edited
  const activity = activities.find((a) => a.id === Number(activityId));

  const [formData, setFormData] = useState({
    name: activity?.name || "",
    category: activity?.category || "",
    location: activity?.location || "",
    description: activity?.description || "",
    scheduledTime: activity?.scheduledTime
      ? new Date(activity.scheduledTime).toISOString().split("T")[0]
      : undefined,
  });

  const [isSubmitting, setIsSubmitting] = useState(false);
  const [formErrors, setFormErrors] = useState([]);

  /** Handles input changes */
  function handleChange(evt) {
    const { name, value } = evt.target;
    setFormData((data) => ({
      ...data,
      [name]: value,
    }));
  }

  async function handleSubmit(evt) {
    evt.preventDefault();
    setIsSubmitting(true);
    setFormErrors([]);
    const formattedScheduledTime = formData.scheduledTime
      ? new Date(formData.scheduledTime).toISOString()
      : undefined;

    const activityToSubmit = {
      ...formData,
      scheduledTime: formattedScheduledTime,
    };

    try {
      await editActivity(activity.id, activityToSubmit);
      navigate(`/trips/${activity.tripId}`);
    } catch (err) {
      setFormErrors(["Error updating activity. Please try again."]);
    } finally {
      setIsSubmitting(false);
    }
  }

  /** Handle delete */
  async function handleDelete() {
    if (window.confirm("Are you sure you want to delete this activity?")) {
      await deleteActivity(activity.id);
      navigate(`/trips/${activity.tripId}`);
    }
  }

  return (
    <div className="h-[calc(100vh-4rem)] flex items-center justify-center px-4 py-10">
      <div className="w-full max-w-xl bg-white p-8 rounded-xl shadow-lg">
        <h1 className="text-3xl font-bold text-gray-600 mb-4">Edit Activity</h1>

        <form onSubmit={handleSubmit} className="space-y-3">
          {/* error alerts */}
          <Alert type="error" messages={formErrors} />

          {/* name */}
          <div>
            <label className="block text-zinc-700 font-medium mb-1" htmlFor="name">
              Activity Name
            </label>
            <input
              type="text"
              name="name"
              id="name"
              value={formData.name}
              onChange={handleChange}
              className="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
              required
            />
          </div>

          {/* category */}
          <div>
            <label className="block text-zinc-700 font-medium mb-1" htmlFor="category">
              Category
            </label>
            <select
              name="category"
              id="category"
              value={formData.category}
              onChange={handleChange}
              className="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
              required
            >
              {validCategories.map((cat) => (
                <option key={cat} value={cat}>
                  {cat.charAt(0).toUpperCase() + cat.slice(1)}
                </option>
              ))}
            </select>
          </div>

          {/* location */}
          <div>
            <label className="block text-zinc-700 font-medium mb-1" htmlFor="location">
              Location
            </label>
            <input
              type="text"
              id="location"
              name="location"
              value={formData.location}
              onChange={handleChange}
              className="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            />
          </div>

          {/* description */}
          <div>
            <label className="block text-zinc-700 font-medium mb-1" htmlFor="description">
              Description
            </label>
            <textarea
              name="description"
              id="description"
              value={formData.description}
              onChange={handleChange}
              className="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            />
          </div>

          {/* scheduled time */}
          <div>
            <label className="block text-zinc-700 font-medium mb-1" htmlFor="scheduledTime">
              Scheduled Time
            </label>
            <input
              type="datetime-local"
              name="scheduledTime"
              id="scheduledTime"
              value={formData.scheduledTime || ""}
              onChange={handleChange}
              className="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            />
          </div>

          <div className="flex space-x-4">
            <button
              type="submit"
              className="w-full bg-blue-500 text-white py-2 rounded-lg hover:bg-blue-600 transition"
              disabled={isSubmitting}
            >
              {isSubmitting ? "Updating..." : "Save Changes"}
            </button>
            <button
              type="button"
              className="w-full bg-gray-400 text-white py-2 rounded-lg hover:bg-gray-500 transition"
              onClick={() => navigate(`/trips/${activity.tripId}`)}
            >
              Cancel
            </button>
          </div>

          <button
            type="button"
            className="w-full bg-red-500 text-white py-2 rounded-lg hover:bg-red-600 transition"
            onClick={handleDelete}
          >
            Delete Activity
          </button>
        </form>
      </div>
    </div>
  );
}

export default EditActivityForm;
