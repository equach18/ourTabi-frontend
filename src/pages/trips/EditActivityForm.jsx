import { useState, useContext } from "react";
import { useNavigate, useParams } from "react-router-dom";
import TripContext from "../../context/TripContext";

function EditActivityForm() {
  const { activities, deleteActivity, editActivity } = useContext(TripContext);
  const { activityId } = useParams(); // Get activity ID from URL params
  const navigate = useNavigate();

  const validCategories = [
    "food",
    "hiking",
    "tours",
    "shopping",
    "adventure",
    "outdoors",
    "other",
  ];

  // Find the activity being edited
  const activity = activities.find((a) => a.id === Number(activityId));

  const [formData, setFormData] = useState({
    name: activity?.name || "",
    category: activity?.category || "",
    location: activity?.location || "",
    description: activity?.description || "",
    scheduledTime: activity?.scheduledTime
      ? new Date(activity.scheduledTime).toISOString().split("T")[0]
      : undefined,
  });

  const [isSubmitting, setIsSubmitting] = useState(false);
  const [formErrors, setFormErrors] = useState([]);

  /** Handles input changes */
  function handleChange(evt) {
    const { name, value } = evt.target;
    setFormData((data) => ({
      ...data,
      [name]: value,
    }));
  }

  async function handleSubmit(evt) {
    evt.preventDefault();
    setIsSubmitting(true);
    setFormErrors([]);

    try {
      await editActivity(activity.id, formData);
      navigate(`/trips/${activity.tripId}`);
    } catch (err) {
      setFormErrors(["Error updating activity. Please try again."]);
    } finally {
      setIsSubmitting(false);
    }
  }

  /** Handle delete */
  async function handleDelete() {
    if (window.confirm("Are you sure you want to delete this activity?")) {
      await deleteActivity(activity.id);
      navigate(`/trips/${activity.tripId}`);
    }
  }

  return (
    <div className="max-w-lg mx-auto mt-10 p-6 bg-white shadow-lg rounded-lg">
      <h1 className="text-2xl font-bold text-blue-500">Edit Activity</h1>

      <form onSubmit={handleSubmit} className="mt-4">
        {/* Name */}
        <div className="mb-4">
          <label className="block text-gray-700">Activity Name</label>
          <input
            type="text"
            name="name"
            value={formData.name}
            onChange={handleChange}
            className="w-full p-2 border rounded-lg"
            required
          />
        </div>

        {/* Category */}
        <div className="mb-4">
          <label className="block text-gray-700">Category</label>
          <select
            type="text"
            name="category"
            value={formData.category}
            onChange={handleChange}
            className="w-full p-2 border rounded-lg"
            required
          >
            {validCategories.map((cat) => (
              <option key={cat} value={cat}>
                {cat.charAt(0).toUpperCase() + cat.slice(1)}
              </option>
            ))}
          </select>
        </div>

        {/* Location */}
        <div className="mb-4">
          <label className="block text-gray-700">Location</label>
          <input
            type="text"
            name="location"
            value={formData.location}
            onChange={handleChange}
            className="w-full p-2 border rounded-lg"
          />
        </div>

        {/* Description */}
        <div className="mb-4">
          <label className="block text-gray-700">Description</label>
          <textarea
            name="description"
            value={formData.description}
            onChange={handleChange}
            className="w-full p-2 border rounded-lg"
          />
        </div>

        {/* Scheduled Time */}
        <div className="mb-4">
          <label className="block text-gray-700">Scheduled Time</label>
          <input
            type="date"
            name="scheduledTime"
            value={formData.scheduledTime || ""}
            onChange={handleChange}
            className="w-full p-2 border rounded-lg"
          />
        </div>

        {/* Error Messages */}
        {formErrors.length > 0 && (
          <div className="text-red-500">
            {formErrors.map((err, i) => (
              <p key={i}>{err}</p>
            ))}
          </div>
        )}

        {/* Submit & Cancel Buttons */}
        <div className="flex space-x-4">
          <button
            type="submit"
            className="bg-blue-500 text-white px-6 py-3 rounded-lg hover:bg-blue-600 w-full"
            disabled={isSubmitting}
          >
            {isSubmitting ? "Updating..." : "Save Changes"}
          </button>
          <button
            type="button"
            className="bg-gray-400 text-white px-6 py-3 rounded-lg w-full"
            onClick={() => navigate(`/trips/${activity.tripId}`)}
          >
            Cancel
          </button>
        </div>

        <button
          type="button"
          className="mt-4 bg-red-500 text-white px-6 py-3 rounded-lg w-full"
          onClick={handleDelete}
        >
          Delete Activity
        </button>
      </form>
    </div>
  );
}

export default EditActivityForm;
