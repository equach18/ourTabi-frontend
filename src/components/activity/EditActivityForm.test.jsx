import { fireEvent, screen, waitFor } from "@testing-library/react";
import { Routes, Route } from "react-router-dom";
import EditActivityForm from "./EditActivityForm";
import {
  renderWithAllProviders,
  testActivity,
} from "../../test-helpers/_testCommons";
const mockNavigate = vi.fn();

vi.mock("react-router-dom", async () => {
  const actual = await vi.importActual("react-router-dom");
  return {
    ...actual,
    useNavigate: () => mockNavigate,
  };
});

describe("EditActivityForm", () => {
  const mockEditActivity = vi.fn();
  const mockDeleteActivity = vi.fn();

  const tripContextValue = {
    activities: [testActivity],
    editActivity: mockEditActivity,
    deleteActivity: mockDeleteActivity,
  };

  function renderForm() {
    return renderWithAllProviders(
      <Routes>
        <Route
          path="/trips/:tripId/activities/:activityId/edit"
          element={<EditActivityForm />}
        />
      </Routes>,
      {
        tripContext: tripContextValue,
        route: `/trips/${testActivity.tripId}/activities/${testActivity.id}/edit`,
      }
    );
  }

  beforeEach(() => {
    mockEditActivity.mockClear();
    mockDeleteActivity.mockClear();
  });

  it("matches snapshot", () => {
    const { asFragment } = renderForm();
    expect(asFragment()).toMatchSnapshot();
  });

  it("renders form with pre-filled data", () => {
    renderForm();

    expect(screen.getByLabelText(/activity name/i)).toHaveValue("Kayaking");
    expect(screen.getByLabelText(/location/i)).toHaveValue("Chicago");
    expect(screen.getByLabelText(/description/i)).toHaveValue("So much fun");
  });

  it("editActivity is called when submitted", async () => {
    const { getByLabelText, getByRole } = renderForm();

    fireEvent.change(getByLabelText(/activity name/i), {
      target: { value: "new activity name" },
    });

    fireEvent.click(getByRole("button", { name: /save changes/i }));

    await waitFor(() => {
      expect(mockEditActivity).toHaveBeenCalledWith(
        testActivity.id,
        expect.objectContaining({ name: "new activity name" })
      );
    });
  });

  it("deleteActivity is called when delete button is clicked", async () => {
    vi.spyOn(window, "confirm").mockReturnValue(true);

    const { getByRole } = renderForm();

    fireEvent.click(getByRole("button", { name: /delete activity/i }));

    await waitFor(() => {
      expect(mockDeleteActivity).toHaveBeenCalledWith(testActivity.id);
    });
  });

  it("navigates to trip detail page after sucessful submission", async () => {
    mockEditActivity.mockResolvedValueOnce();
    renderForm();

    fireEvent.change(screen.getByLabelText(/activity name/i), {
      target: { value: "Updated" },
    });

    fireEvent.click(screen.getByRole("button", { name: /save changes/i }));

    await waitFor(() => {
      expect(mockEditActivity).toHaveBeenCalledWith(
        testActivity.id,
        expect.objectContaining({ name: "Updated" })
      );
      expect(mockNavigate).toHaveBeenCalledWith(`/trips/${testActivity.tripId}`);
    });
  });
});
