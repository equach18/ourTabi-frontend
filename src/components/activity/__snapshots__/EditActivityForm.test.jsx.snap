// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`EditActivityForm > matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="h-[calc(100vh-4rem)] flex items-center justify-center px-4 py-10"
  >
    <div
      class="w-full max-w-xl bg-white p-8 rounded-xl shadow-lg"
    >
      <h1
        class="text-3xl font-bold text-gray-600 mb-4"
      >
        Edit Activity
      </h1>
      <form
        class="space-y-3"
      >
        <div>
          <label
            class="block text-zinc-700 font-medium mb-1"
            for="name"
          >
            Activity Name
          </label>
          <input
            class="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            id="name"
            name="name"
            required=""
            type="text"
            value="Kayaking"
          />
        </div>
        <div>
          <label
            class="block text-zinc-700 font-medium mb-1"
            for="category"
          >
            Category
          </label>
          <select
            class="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            id="category"
            name="category"
            required=""
          >
            <option
              value="food"
            >
              Food
            </option>
            <option
              value="hiking"
            >
              Hiking
            </option>
            <option
              value="tours"
            >
              Tours
            </option>
            <option
              value="shopping"
            >
              Shopping
            </option>
            <option
              value="adventure"
            >
              Adventure
            </option>
            <option
              value="outdoors"
            >
              Outdoors
            </option>
            <option
              value="other"
            >
              Other
            </option>
          </select>
        </div>
        <div>
          <label
            class="block text-zinc-700 font-medium mb-1"
            for="location"
          >
            Location
          </label>
          <input
            class="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            id="location"
            name="location"
            type="text"
            value="Chicago"
          />
        </div>
        <div>
          <label
            class="block text-zinc-700 font-medium mb-1"
            for="description"
          >
            Description
          </label>
          <textarea
            class="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            id="description"
            name="description"
          >
            So much fun
          </textarea>
        </div>
        <div>
          <label
            class="block text-zinc-700 font-medium mb-1"
            for="scheduledTime"
          >
            Scheduled Time
          </label>
          <input
            class="w-full py-1 px-3 border border-stone-300 rounded-md focus:outline-none focus:ring-2 focus:ring-emerald-500"
            id="scheduledTime"
            name="scheduledTime"
            type="datetime-local"
            value="2025-07-10"
          />
        </div>
        <div
          class="flex space-x-4"
        >
          <button
            class="w-full bg-blue-500 text-white py-2 rounded-lg hover:bg-blue-600 transition"
            type="submit"
          >
            Save Changes
          </button>
          <button
            class="w-full bg-gray-400 text-white py-2 rounded-lg hover:bg-gray-500 transition"
            type="button"
          >
            Cancel
          </button>
        </div>
        <button
          class="w-full bg-red-500 text-white py-2 rounded-lg hover:bg-red-600 transition"
          type="button"
        >
          Delete Activity
        </button>
      </form>
    </div>
  </div>
</DocumentFragment>
`;
